### ISSUE TITLE
[SWMR] can't open for append if file is being read by another process

### ISSUE THREAD

**Issue by cpascual** [h5py/h5py#712]
Hi, 

I want to do the following:
- open a file for write (create a file and create a dataset in it)
- close the file
- open the file for read in an independent process that keeps it open
- open the file in 'a' mode and extend the existing dataset

According to section 5.1 of [the hdf5 SWMR docs](https://www.hdfgroup.org/HDF5/docNewFeatures/SWMR/HDF5_SWMR_Users_Guide.pdf) this should be possible. Note that it explicitly states:

> after step 1, the file can be opened by the writer and reader(s) in any order

But when I try (using, e.g. [this code](https://gist.github.com/cpascual/026f115091e7fd45057f10b279110ce7) with h5py 2.6  and hdf5 v 1.10.0) it fails while trying to open in 'a' mode:

```
hdf5_version= 1.10.0
READER: Open file for read
Traceback (most recent call last):
  File "openbug2.py", line 48, in <module>
    open_for_append()
  File "openbug2.py", line 36, in open_for_append
    f = h5py.File(FNAME, 'a', libver='latest')
  File "/home/cpascual/.local/devenv/lib/python/site-packages/h5py-2.6.0-py2.7-linux-x86_64.egg/h5py/_hl/files.py", line 272, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
  File "/home/cpascual/.local/devenv/lib/python/site-packages/h5py-2.6.0-py2.7-linux-x86_64.egg/h5py/_hl/files.py", line 106, in make_fid
    fid = h5f.create(name, h5f.ACC_EXCL, fapl=fapl, fcpl=fcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/nfs/home/cpascual/src/h5py/h5py/_objects.c:2582)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/nfs/home/cpascual/src/h5py/h5py/_objects.c:2541)
  File "h5py/h5f.pyx", line 96, in h5py.h5f.create (/nfs/home/cpascual/src/h5py/h5py/h5f.c:1997)
IOError: Unable to create file (Unable to open file: name = 'foo.h5', errno = 17, error message = 'file exists', flags = 15, o_flags = c2)

```

Also, incidentally, note that the same code works if using hdf v 1.8.16 (which implies that existing code may break, even if no SWMR feature is used at all).

**Comment by stevendbrown** [comment]
i am having the same problem running h5py 2.6.0, HDF5 1.10.0, and python 2.7.10 on Ubuntu Linux 15.10.

**Comment by epourmal** [comment]
Steven,

In the current implementation a reader should come after the writer. If a reader opens a file only another reader can open the file; no writer can open it.

Elena

Sent from my iPhone

On Jun 1, 2016, at 6:09 PM, Steven Brown <notifications@github.com<mailto:notifications@github.com>> wrote:


i am having the same problem running h5py 2.6.0, HDF5 1.10.0, and python 2.7.10 on Ubuntu Linux 15.10.

-

**Comment by stevendbrown** [comment]
hi Elena,

this behavior appears to differ from the HDF5 specification (section 5.1 of the user manual as cited by @cpascual). am i reading or interpreting the HDF5 spec wrong?

thanks,
steve

**Comment by epourmal** [comment]
Hi Steve,

Sorry! Our fault… I am afraid we have an outdated document on the Web (https://www.hdfgroup.org/HDF5/docNewFeatures/SWMR/HDF5_SWMR_Users_Guide.pdf).

When we introduced file locking, the SWMR programming model changed (and it happened very close to the 1.10.0 release).

When opening a file, a writer needs to place an exclusive lock for the time it modifies the superblock to indicate that the file will be under SWMR access. A reader acquires a shared lock. When the reader comes first, the writer cannot open the file anymore. This is why you see the failure.

Once again I apologize for the oversight, we will fix it ASAP.

Thank you for finding the error!

Elena
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Elena Pourmal  The HDF Group  http://hdfgroup.org
1800 So. Oak St., Suite 203, Champaign IL 61820
217.531.6112
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~




On Jun 1, 2016, at 9:35 PM, Steven Brown <notifications@github.com<mailto:notifications@github.com>> wrote:


hi Elena,

this behavior appears to differ from the HDF5 specification (section 5.1 of the user manual as cited by @cpascual<https://github.com/cpascual>). am i reading or interpreting the HDF5 spec wrong?

thanks,
steve

**Comment by cpascual** [comment]
I think this defeats one of the main advantages of the SWMR feature (which is not requiring coordination between readers and writer).

**Comment by cpascual** [comment]
Also, IMHO this is a **regression** which may break code that worked in 1.8. 

Consider the following scenario:

A scientific data acquisition program writes on an hdf5 file. On each experiment, it opens the file, writes the acquired data and closes the file. Related experiments use different datagroups **on the same file**. Now, if a user is viewing the data from a previous experiment with a reader, the data acquisition will fail to re-open the file .

Note that the above happens even when no SWMR mode is used at all, while in v1.8, it was possible to work in the mentioned scenario (admittedly, risking to **read** inconsistent data, but at least the writer was not blocked and the reader could re-read later-on)

**Comment by epourmal** [comment]
On Jun 2, 2016, at 1:18 PM, cpascual <notifications@github.com<mailto:notifications@github.com>> wrote:


Also, IMHO this is a regression which may break code that worked in 1.8.

Totally agree. We are reconsidering the locking issue. There will be RFC (Request for Comments) coming to HDF-FORUM. Stay tuned.

Elena

Consider the following scenario:

A scientific data acquisition program writes on an hdf5 file. On each experiment, it opens the file, writes the acquired data and closes the file. Related experiments use different datagroups on the same file. Now, if a user is viewing the data from a previous experiment with a reader, the data acquisition will fail to re-open the file .

Note that the above happens even when no SWMR mode is used at all, while in v1.8, it was possible to work in the mentioned scenario (admittedly, risking to read inconsistent data, but at least the writer was not blocked and the reader could re-read later-on)

**Comment by epourmal** [comment]
Although we plan to have a way to turn off file locking (probably in 10.0.1, due mid-summer-ish), this scenario is exactly what the file locking is intended to prevent since this sort of access can result in unpredictable reader crashes.

See the upcoming RFC, though, for more information on the levers we'll provide to users that will allow control of the locking feature.

Dana Robinson
Software Engineer
The HDF Group

**Comment by cpascual** [comment]
@epourmal , I see that hdf1.10.0-patch1 is out, but from the release notes, I gather that it does not address this problem. Any updates / ETA on this topic? 

Cheers, 

Carlos

**Comment by episodeyang** [comment]
I'm running into the same issue.

**Comment by epourmal** [comment]
We hope to address the locking issue in1.10.1 later in the year. We will post the proposed solution on HDF forum. Stay tuned.

**Comment by cpascual** [comment]
Hi,

The [release notes of the recent HDF5 1.10.1](https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.1/src/hdf5-1.10.1-RELEASE.txt) state :

>  Added a mechanism for disabling the SWMR file locking scheme.
      The file locking calls used in HDF5 1.10.0 (including patch1)
      will fail when the underlying file system does not support file
      locking or where locks have been disabled. To disable all file
      locking operations, an environment variable named
      HDF5_USE_FILE_LOCKING can be set to the five-character string
      'FALSE'. This does not fundamentally change HDF5 library
      operation (aside from initial file open/create, SWMR is lock-free),
      but users will have to be more careful about opening files
      to avoid problematic access patterns (i.e.: multiple writers)
      that the file locking was designed to prevent.

But I am not sure what is the exact consequence of this regarding the issue that [I originally reported](https://github.com/h5py/h5py/issues/712#issue-156270678).
Could some HDF5 expert ( @epourmal  ?) clarify whether the access pattern that I mentioned is now possible now *possible* and *safe* with HDF5 1.10.1? 

(I use "*possible*" and "*safe*" in the sense that my access pattern is *not possible** (the example fails) with 1.10.0-patch1 and it is *possible-but-unsafe* with 1.8.16 (the eample runs but 

In other words: by disabling the file locking in 1.10.1 and using SWMR, will I be able to safely append data to an existing file that is already open in read-only mode by a different process?

**Comment by alfonnews** [comment]
I have the same question as cpascual:
using SWMR, will I be able to safely append data to an existing file that is already open in read-only mode by a different process?

Thanks by your time

**Comment by akozlu** [comment]
Running to the same issue, but with a different error trace: 
File "/Users/PycharmProjects/TennisPrediction/Dataset.py", line 10, in __init__
    store = pd.HDFStore('storage.h5', mode='r')
  File "/Users/anaconda3/lib/python3.6/site-packages/pandas/io/pytables.py", line 467, in __init__
    self.open(mode=mode, **kwargs)
  File "/Users/python3.6/site-packages/pandas/io/pytables.py", line 612, in open
    raise IOError(str(e))
OSError: HDF5 error back trace

  File "H5F.c", line 511, in H5Fopen
    unable to open file
  File "H5Fint.c", line 1519, in H5F_open
    unable to lock the file
  File "H5FD.c", line 1650, in H5FD_lock
    driver lock request failed
  File "H5FDsec2.c", line 941, in H5FD_sec2_lock
    unable to lock file, errno = 35, error message = 'Resource temporarily unavailable'

End of HDF5 error back trace

**Comment by aparamon** [comment]
@cpascual @alfonnews 
The updated docs
https://support.hdfgroup.org/HDF5/docNewFeatures/SWMR/HDF5_SWMR_Users_Guide.pdf
https://support.hdfgroup.org/HDF5/docNewFeatures/SWMR/Design-HDF5-FileLocking.pdf
seem to indicate clearly that SWMR writer **must** gain file access before any readers, even with `HDF5_USE_FILE_LOCKING=FALSE`.

I propose that improvement suggestion for lifting this limitation (allowing SWMR readers exist by the time the writer is created) is submitted to https://forum.hdfgroup.org. With the recent announcement of public HDF Group issue tracker and opening up the development process for external contributors, chances are getting higher that this feature might be implemented rather sooner than never.

**Comment by cpascual** [comment]
Hi @aparamon ,thanks for pointing to the new docs. I agree in that they are now clear about the fact that what we intended is not currently allowed.

I just created an entry in the HDF Group forum with the feature request. Let's see if someone gets interested:

https://forum.hdfgroup.org/t/allow-swmr-reader-to-be-set-before-swmr-writer/5244

**Comment by wrobell** [comment]
The current h5py documentation is bit misleading
```
No synchronisation is required between processes...
```
http://docs.h5py.org/en/stable/swmr.html

But you need to coordinate readers to not access existing file before the writer opens it.

**Comment by aparamon** [comment]
Upstream issue:
https://jira.hdfgroup.org/browse/HDFFV-10706

**Comment by NumesSanguis** [comment]
How is this issue going?
I have little control over when readers access the HDF5 file (multi-user, pipeline environment), and as long as a reader in a different processes has access to it, I can't change the mode to `.swmr_mode = False`. Therefore, I cannot create safely any new dataset at a later stage (what datasets will be in there is not known at creation time).

Would it be possible for a writer to forcefully change `.swmr_mode = False`, even if that will block existing readers? That would be better than be stuck with a HDF5 where no dataset can be created inside anymore.

@aparamon Your linked upstream issue is not accessible without logging in.

**Comment by takluyver** [comment]
No HDF5 issues can be accessed without logging in - that's the way their bugtracker works. It should be possible for anyone who wants to create an account - I'm guessing it's via this page:
https://www.hdfgroup.org/register/

I'm not sure about that, though - I think it was different when I created my account.

The upstream issue about this is still open, with no recent updates.

**Comment by vasole** [comment]
Just to comment that we are not using the SWMR but we have no problems opening an HDF5 file for writing while open for reading by other process provided the environment variable HDF5_USE_FILE_LOCKING is set to FALSE at least for all the reader processes.

**Comment by takluyver** [comment]
It looks like the `HDF5_USE_FILE_LOCKING` environment variable is meant as an escape hatch in case you have a filesystem that doesn't support locks. The [release notes for HDF5 1.10.1](https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.1/src/hdf5-1.10.1-RELEASE.txt) say:

> The file locking calls used in HDF5 1.10.0 (including patch1)  will fail when the underlying file system does not support file locking or where locks have been disabled. To disable all file locking operations, an environment variable named HDF5_USE_FILE_LOCKING can be set to the five-character string 'FALSE'. This does not fundamentally change HDF5 library operation (aside from initial file open/create, SWMR is lock-free), but users will have to be more careful about opening files to avoid problematic access patterns (i.e.: multiple writers) that the file locking was designed to prevent.

By my reading, that means you can disable locking, but then it's entirely up to you to avoid the kind of problems that the locks would prevent. Of course, those problems might be very rare in a given usage pattern.

**Comment by vasole** [comment]
The problem is that if you do not set HDF5_USE_FILE_LOCKING to FALSE for the readers you are going to block any writer. Therefore my comment. What we do is that the (SINGLE!) writer process does not get the environment variable set to FALSE while the readers get it. In fact, pure readers (our diagnostics applications) set the environment variable themselves prior to import h5py to make sure they are not going to interfere with the single writer (our acquisition system).

**Comment by vjp23** [comment]
Any updates on this issue?

**Comment by takluyver** [comment]
I'm going to close this because - as far as I can see, any issues or limitations here are with HDF5 itself rather than h5py - although PRs are welcome to improve h5py's documentation around SWMR.

As I understand it, the status is:

- SWMR writers must open the file and switch it to SWMR mode before readers open it.
- You can bypass the file locking by setting the environment variable `HDF5_USE_FILE_LOCKING=FALSE`, but if you use that to do anything that would be prevented by locking, this is strictly unsafe (although in some use cases, it might just happen to work).
